### PR TITLE
Atomic add and sub for int32, int64, float32, float64

### PR DESCRIPTION
**Pull request by reazulhoque**
This PR will add support for atomic_add and atomic_sub for int32, int64, float32 and float64 datatypes and will allow extension for other operations like multiply in the future as well.

### CODE REVIEW

File: numba/dppl/tests/dppl/test_atomic_op.py [hunk 1]
```diff
@@ -87,6 +161,91 @@ def test_atomic_add1(self):
...
         self.assertTrue(np.all(ary == gold))
 
+
+    def test_atomic_sub_local_int32(self):
+        ary = np.random.randint(0, 32, size=32).astype(np.uint32)
+        orig = ary.copy()
+
+        dppl_atomic_sub = dppl.kernel('void(uint32[:])')(atomic_sub_int32)
+        with ocldrv.igpu_context(0) as device_env:
+            dppl_atomic_sub[32, dppl.DEFAULT_LOCAL_SIZE](ary)
```
Comment (PokhodenkoSA) [inline numba/dppl/tests/dppl/test_atomic_op.py:174]: If local size is default why to set it explicitly? What if you will omit it?

File: setup.py [hunk 2]
```diff
@@ -213,6 +213,7 @@ def check_file_at_path(path2file):
         else:
             omplinkflags = ['-fopenmp']
 
+    tbb_root = None
```
Comment (PokhodenkoSA) [inline setup.py]: Do not eliminate tbb by yourself. We know that Numba 0.49.1 and 0.50.1 does not work with TBB beta08. Just do not install tbb in your conda environment.

File: numba/dppl/tests/dppl/test_atomic_op.py (hunk 1 above)
Comment (reazulhoque) [reply numba/dppl/tests/dppl/test_atomic_op.py:174]: The way we are able to pass the global and local size is by using the __getitem__ attribute. Now, in the __getitem__() attribute, we receive these arguments using the parameter args (https://github.com/IntelPython/numba/blob/pydppl/numba/dppl/compiler.py#L304)

It behaves in the following way:

Case 1:
```
global_size = 32
dppl_atomic_sub[global_size]
```

If we test args in __getitem__() we will find it to be a single integer

Case 2:
```
global_size = (32, 32)
dppl_atomic_sub[global_size]
```
If we test args in __getitem__() we will find it to be a tuple of two integers (32, 32)

Case 3:
```
global_size = 32
local_size = 8
dppl_atomic_sub[global_size, local_size]
```
If we test args in __getitem__() we will find it to be a tuple of two integers (32, 8)

**Issue**: we will find identical type for both case 2 and 3 and therefore can not distinguish between which the user intended as global size and which as local size

Solution 1: Users always wrap whatever size they are passing in a list or tuple

So,

Case 1 becomes:
```
global_size = 32
dppl_atomic_sub[[global_size]] / dppl_atomic_sub[(global_size)] 
```

Users are always forced to wrap parameters as a list even for single value.

Solution 2: Always send two parameters:
Case 1 becomes:
```
global_size = 32
dppl_atomic_sub[global_size, some_default_value_if_user_don't_want_to_set_it]
```

I have chosen soultion 2. We can always discuss which is better or if we need to completely change how we allow users to specify the global and local size when invoking a kernel.

I hope this answers the question.

File: setup.py (hunk 2 above)
Comment (reazulhoque) [reply setup.py]: I agree, I will revert it. Thanks for catching it.

File: numba/dppl/compiler.py [hunk 3]
```diff
@@ -118,7 +119,8 @@ def compile_kernel(device, pyfunc, args, access_types, debug=False):
     cres = compile_with_dppl(pyfunc, None, args, debug=debug)
     func = cres.library.get_function(cres.fndesc.llvm_func_name)
     kernel = cres.target_context.prepare_ocl_kernel(func, cres.signature.args)
-    oclkern = DPPLKernel(device_env=device,
+    oclkern = DPPLKernel(context=cres.target_context,
```
Comment (diptorupd) [inline numba/dppl/compiler.py:131]: A comment explaining why the "context" param was added will be good.

File: numba/dppl/ocl/atomics/atomic_ops.cl [hunk 4]
```diff
@@ -0,0 +1,216 @@
+#ifdef cl_khr_int64_base_atomics
```
Comment (diptorupd) [inline numba/dppl/ocl/atomics/atomic_ops.cl:5]: Some comments should be added. If we are replicating what llvm-spirv does then a link to upstream will be helpful.

File: numba/dppl/ocl/oclimpl.py [hunk 5]
```diff
@@ -161,12 +164,100 @@ def sub_group_barrier_impl(context, builder, sig, args):
...
+        ll_val = ir.IntType(32)
+        ll_p = ll_val.as_pointer()
+        if fn_type == "add":
+            name = "numba_dppl_atomic_add_i32"
+        elif fn_type == "sub":
+            name = "numba_dppl_atomic_sub_i32"
+        else:
+            raise TypeError("Operation type is not supported %s" %
+                             (fn_type))
+
```
Comment (diptorupd) [inline numba/dppl/ocl/oclimpl.py]: Why extra newlines?

File: numba/dppl/ocl/oclimpl.py [hunk 6]
```diff
@@ -250,20 +353,16 @@ def _generic_array(context, builder, shape, dtype, symbol_name, addrspace):
...
 
-        # Convert to generic address-space
-        dataptr = context.addrspacecast(builder, gvmem,
-                                        target.SPIR_GENERIC_ADDRSPACE)
-
     else:
         raise NotImplementedError("addrspace {addrspace}".format(**locals()))
 
-    return _make_array(context, builder, dataptr, dtype, shape)
+    return _make_array(context, builder, gvmem, dtype, shape, addrspace=addrspace)
```
Comment (diptorupd) [inline numba/dppl/ocl/oclimpl.py:374]: I know why the change was done, since you explained them to me 👍. But, please add a comment for others.

File: numba/dppl/target.py [hunk 7]
```diff
@@ -51,6 +51,9 @@ def load_additional_registries(self):
 SPIR_VERSION = (2, 0)
 
 
+LINK_ATOMIC = 111
```
Comment (diptorupd) [inline numba/dppl/target.py:55]: What is this?

**Review by diptorupd — COMMENTED** [review]
Overall looks good. Please add comments in the few places I mentioned and also overall documentation can be improved. Apart from that ready to go 👍. Thanks!

File: numba/dppl/target.py (hunk 7 above)
Comment (reazulhoque) [reply numba/dppl/target.py:55]: Option was added to dppl target context to identify if we need to link our kernel with any other implementation, e.g. **only kernels** containing atomics should be linked with spirv binary containing the implementation of those atomic operations. The options are passed in form of dictionary where the keys define which object files need to be linked. Here, we are defining that "key" for linking the atomic.

File: numba/dppl/ocl/atomics/atomic_ops.cl (hunk 4 above)
Comment (reazulhoque) [reply numba/dppl/ocl/atomics/atomic_ops.cl:5]: Comment added

File: numba/dppl/ocl/oclimpl.py (hunk 6 above)
Comment (reazulhoque) [reply numba/dppl/ocl/oclimpl.py:374]: Comment added.

File: numba/dppl/compiler.py (hunk 3 above)
Comment (reazulhoque) [reply numba/dppl/compiler.py:131]: Comment added.

File: numba/dppl/ocl/atomics/atomic_ops.cl [hunk 8]
```diff
@@ -0,0 +1,220 @@
+/* OpenCL extension specification states extensions should define the extension
+   and to use it we need to enable is the following way using the pragma.
+   Link to specification: https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_Ext.html
+*/
+#ifdef cl_khr_int64_base_atomics
+  #pragma OPENCL EXTENSION cl_khr_int64_base_atomics: enable
+
+  long numba_dppl_atomic_add_i64_local(volatile __generic long *p, long val) {
```
Comment (AlexanderKalistratov) [inline numba/dppl/ocl/atomics/atomic_ops.cl:8]: Why we are not using built-in atomic_add?

File: numba/dppl/ocl/oclimpl.py [hunk 9]
```diff
@@ -161,12 +164,98 @@ def sub_group_barrier_impl(context, builder, sig, args):
     return _void_value
 
 
+def insert_and_call_atomic_fn(context, builder, sig, fn_type,
+                              dtype, ptr, val, addrspace):
+    ll_p = None
+    name = ""
+    if dtype.name == "int32" or dtype.name == "uint32":
+        ll_val = ir.IntType(32)
```
Comment (AlexanderKalistratov) [inline numba/dppl/ocl/oclimpl.py:169]: Why no check for support 32-bit atomics?

File: numba/dppl/spirv_generator.py [hunk 10]
```diff
@@ -116,6 +124,17 @@ def finalize(self):
         spirv_path = self._track_temp_file("generated-spirv")
         self._cmd.generate(ipath=self._llvmfile, opath=spirv_path)
 
+        binary_paths = [spirv_path]
+        for key in list(self.context.link_binaries.keys()):
+            del self.context.link_binaries[key]
```
Comment (AlexanderKalistratov) [inline numba/dppl/spirv_generator.py:129]: Why we are deleting this keys?

File: numba/dppl/spirv_generator.py [hunk 11]
```diff
@@ -116,6 +124,17 @@ def finalize(self):
...
+            if key == LINK_ATOMIC:
```
Comment (AlexanderKalistratov) [inline numba/dppl/spirv_generator.py:130]: We are not checking for value, only for keys, so probably, it would be better to use `set` instead of `dict`? Or you are planning to use value too? If so it is better to add check for value here

File: numba/dppl/target.py (hunk 7 above)
Comment (AlexanderKalistratov) [reply numba/dppl/target.py:55]: Why not to use string as key?
Something like that:
```python
LINK_ATOMIC= 'PYDPPL_LINK_ATOMIC_KEY'
```
?

File: numba/dppl/ocl/atomics/atomic_ops.cl (hunk 8 above)
Comment (reazulhoque) [reply numba/dppl/ocl/atomics/atomic_ops.cl:8]: I am not aware of built-in atomic add in opencl 2. I might have missed it. Also, I wanted to have consistent compare and swap implementation for all atomic operations.

File: numba/dppl/ocl/oclimpl.py (hunk 9 above)
Comment (reazulhoque) [reply numba/dppl/ocl/oclimpl.py:169]: 32-bit atomics are not part of any extensiona and comes as default, that is why I did not put in any check. Do you think I have wrong assumptions here?

File: numba/dppl/spirv_generator.py (hunk 10 above)
Comment (reazulhoque) [reply numba/dppl/spirv_generator.py:129]: I wanted to make sure we only have this option enabled for kernels using atomics. To make sure in the cases where we have multiple kernels and only the first one uses atomics, we do not want to link the later kernels with the atomic implementation.

File: numba/dppl/spirv_generator.py (hunk 11 above)
Comment (reazulhoque) [reply numba/dppl/spirv_generator.py:130]: I wanted to have the option of being able to use it later, as far as I know set is implemented using a hashmap so we will not be saving much here (https://stackoverflow.com/questions/3949310/how-is-set-implemented). Please let me know what you think.

File: numba/dppl/ocl/oclimpl.py (hunk 9 above)
Comment (AlexanderKalistratov) [reply numba/dppl/ocl/oclimpl.py:169]: OpenCL 2.0 indeed doesn't specify any extensions for 32-bit atomics:
https://www.khronos.org/registry/OpenCL//sdk/2.0/docs/man/xhtml/

While OpenCL 2.1 do specifies the following extensions for 32-bit atomics:
[cl_khr_global_int32_base_atomics](https://www.khronos.org/registry/OpenCL//sdk/2.1/docs/man/xhtml/cl_khr_global_int32_base_atomics.html)
[cl_khr_global_int32_extended_atomics](https://www.khronos.org/registry/OpenCL//sdk/2.1/docs/man/xhtml/cl_khr_global_int32_extended_atomics.html)
[cl_khr_local_int32_base_atomics](https://www.khronos.org/registry/OpenCL//sdk/2.1/docs/man/xhtml/cl_khr_local_int32_base_atomics.html)
[cl_khr_local_int32_extended_atomics](https://www.khronos.org/registry/OpenCL//sdk/2.1/docs/man/xhtml/cl_khr_local_int32_extended_atomics.html)


https://www.khronos.org/registry/OpenCL//sdk/2.1/docs/man/xhtml/

File: numba/dppl/spirv_generator.py (hunk 11 above)
Comment (AlexanderKalistratov) [reply numba/dppl/spirv_generator.py:130]: Performance is not what I'm worriyng about.
It's more about case, when someone would wrote code like:
```python
link_libraries[LINK_ATOMIC] = False
```

But it would still link with the library

File: numba/dppl/ocl/atomics/atomic_ops.cl (hunk 8 above)
Comment (AlexanderKalistratov) [reply numba/dppl/ocl/atomics/atomic_ops.cl:8]: OpenCL 2.0 specifies the following basic 64-bit atomic operations:

Function | Description
-- | --
atom_add | Compute (old + val)
atom_sub | Compute (old - val)
atom_xchg | Swaps the 32-bit value (old) stored at location p with new value val.
atom_inc | Compute (old + 1)
atom_dec | Compute (old - 1)
atom_cmpxchg | Compute (old == cmp) ? val : old

https://www.khronos.org/registry/OpenCL//sdk/2.0/docs/man/xhtml/cl_khr_int64_base_atomics.html

I believe it is true for 32-bit atomics too. 
And for OpenCL 2.1 too.

Though, taking into consideration what we are going to get rid of OpenCL implementation we could left it as is.

Comment (reazulhoque) [reply numba/dppl/ocl/atomics/atomic_ops.cl:8]: I agree.

File: numba/dppl/ocl/oclimpl.py (hunk 9 above)
Comment (reazulhoque) [reply numba/dppl/ocl/oclimpl.py:169]: It mentions that the local int32 base atomics are core now. https://www.khronos.org/registry/OpenCL//sdk/2.1/docs/man/xhtml/. Also atomic compare and exchange is being used here so only 64 bit atomics are coming as part of extension.

File: numba/dppl/spirv_generator.py (hunk 11 above)
Comment (reazulhoque) [reply numba/dppl/spirv_generator.py:130]: I agree in general but in this case the future use case is not True or False (at least not now). Currently the key being there indicates we want to link which is why we are also removing the key as soon as we are done.

File: numba/dppl/target.py (hunk 7 above)
Comment (reazulhoque) [reply numba/dppl/target.py:55]: Simply beacause I prefer avoiding string whenever I can. I understand it will be hashed and might not make a big difference but I still prefer using integer.

**Comment by adarshyoga** [comment]
Are there any changes that are required to this PR? If not, can we have this merged soon?
